File: Bindings/kotlin/RoyalVNCAndroidTest/royalvnc/src/main/java/com/royalapps/royalvnc/VncFrameEncodings.kt
package com.royalapps.royalvnc

import com.sun.jna.*

data class VncFrameEncodings(
    internal val ptr: Pointer /* rvnc_frame_encodings_t */
): AutoCloseable {
    companion object {
        operator fun invoke(): VncFrameEncodings {
            val ptr = RoyalVNCKit.rvnc_frame_encodings_create()

            val frameEncodings = VncFrameEncodings(ptr)

            return frameEncodings
        }

        operator fun invoke(frameEncodingTypes: Array<VncFrameEncodingType>): VncFrameEncodings {
            val ptr = RoyalVNCKit.rvnc_frame_encodings_create()

            val frameEncodings = VncFrameEncodings(ptr)

            for (frameEncodingType in frameEncodingTypes) {
                frameEncodings.appendFrameEncodingType(frameEncodingType)
            }

            return frameEncodings
        }
    }

    private fun appendFrameEncodingType(frameEncodingType: VncFrameEncodingType) {
        when (frameEncodingType) {
            VncFrameEncodingType.RRE
                 -> RoyalVNCKit.rvnc_frame_encodings_append_rre(ptr)

            VncFrameEncodingType.CORRE
                -> RoyalVNCKit.rvnc_frame_encodings_append_corre(ptr)

            VncFrameEncodingType.HEXTILE
                -> RoyalVNCKit.rvnc_frame_encodings_append_hextile(ptr)

            VncFrameEncodingType.ZLIB
                -> RoyalVNCKit.rvnc_frame_encodings_append_zlib(ptr)

            VncFrameEncodingType.TIGHT
                -> RoyalVNCKit.rvnc_frame_encodings_append_tight(ptr)

            VncFrameEncodingType.ZRLE
                -> RoyalVNCKit.rvnc_frame_encodings_append_zrle(ptr)
        }
    }

    override fun close() {
        RoyalVNCKit.rvnc_frame_encodings_destroy(ptr)
    }
}

File: Bindings/dotnet/RoyalApps.RoyalVNCKit/VncFrameEncodings.cs
using System;
using static RoyalApps.RoyalVNCKit.RoyalVNCKit;

namespace RoyalApps.RoyalVNCKit;

internal sealed unsafe class VncFrameEncodings: IDisposable
{
    bool _isDisposed;

    internal void* Instance { get; private set; }

    internal VncFrameEncodings(void* instance)
    {
        ArgumentNullException.ThrowIfNull(instance);
        Instance = instance;
    }

    internal VncFrameEncodings(VncFrameEncodingType[] frameEncodingTypes)
    {
        Instance = rvnc_frame_encodings_create();

        foreach (var frameEncodingType in frameEncodingTypes) {
            AppendFrameEncodingType(frameEncodingType);
        }
    }

    private void AppendFrameEncodingType(VncFrameEncodingType frameEncodingType)
    {
        ArgumentNullException.ThrowIfNull(Instance);
        switch (frameEncodingType) {
            case VncFrameEncodingType.Rre:
                rvnc_frame_encodings_append_rre(Instance);
                break;
            case VncFrameEncodingType.CoRre:
                rvnc_frame_encodings_append_corre(Instance);
                break;
            case VncFrameEncodingType.Hextile:
                rvnc_frame_encodings_append_hextile(Instance);
                break;
            case VncFrameEncodingType.Zlib:
                rvnc_frame_encodings_append_zlib(Instance);
                break;
            case VncFrameEncodingType.Tight:
                rvnc_frame_encodings_append_tight(Instance);
                break;
            case VncFrameEncodingType.Zrle:
                rvnc_frame_encodings_append_zrle(Instance);
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(frameEncodingType), frameEncodingType, $"Unknown {nameof(VncFrameEncodingType)}");
        }
    }

    public void Dispose()
    {
        if (_isDisposed)
            return;

        _isDisposed = true;

        if (Instance is null)
            return;

        rvnc_frame_encodings_destroy(Instance);
        Instance = null;
    }
}
